wp-lazy-loading.php

```diff
@@ -141,14 +139,36 @@ function wp_add_lazy_load_attributes( $content, $context = null ) {
 
 	return preg_replace_callback(
 		'/<img\s[^>]+>/',
-		function( array $matches ) {
+		function( array $matches ) use( $content, $context ) {
 			if ( ! preg_match( '/\sloading\s*=/', $matches[0] ) ) {
-				return str_replace( '<img', '<img loading="lazy"', $matches[0] );
+				$tag_html = $matches[0];
+
+				/**
+				 * Filters the `loading` attribute value. Default `lazy`.
+				 *
+				 * Returning `false` or an empty string will not add the attribute.
+				 * Returning `true` will add the default value.
+				 *
+				 * @since (TBD)
+				 *
+				 * @param string $default The filtered value, defaults to `lazy`.
+				 * @param string $tag_html The tag's HTML.
+				 * @param string $content The HTML containing the image tag.
+				 * @param string $context Optional. Additional context. Defaults to `current_filter()`.
+				 */
+				$value = apply_filters( 'wp_set_image_loading_attr', 'lazy', $tag_html, $content, $context );
```

Comment:
I think passing contextual data here is good, but I'm a bit wary of passing the content here. Anything this could be used for is probably to run another regex, which isn't something we should encourage. It would be great to have a post ID or potentially attachment ID, but since neither is passed to here, and it may be unreliable or expensive to get these, maybe we should just leave it out for now and only pass `$context`.

This is something where maybe we can enhance in the future, but I'd prefer to leave out `$content` for now, since it wouldn't really help much.

Reply:
I wonder whether we should remove the `$context` from the more generic `wp_lazy_loading_enabled` filter and only have it here. It would result in a more clarified distinction between the two filters:
* `wp_lazy_loading_enabled` is to enable/disable lazy-loading globally for specific tags.
* `wp_set_image_loading_attr` allows modifying the `loading` attribute for images specifically on an individual level.

Reply:
> This is something where maybe we can enhance in the future, but I'd prefer to leave out $content for now, since it wouldn't really help much.

Don't think this is a good idea. Having more context in filters/actions makes them more useful. If we remove it, and a plugin has to get the context in order to work, the only option would be to disable core's functionality, copy the code to the plugin and "tweak" it so it works. That brings a lot more problems down the line. Why would we want to "restrict" the context and not support edge cases?

Reply:
This filter isn't specially useful, as not all the tags on put through this function..

Reply:
> I wonder whether we should remove the $context from the more generic wp_lazy_loading_enabled filter...

Same as above. Less context means less useful hooks :) 

> `wp_lazy_loading_enabled` is to enable/disable lazy-loading globally for specific tags.

Not exactly. Currently this is "enable/disable lazy-loading globally for img tags in specific context". It is not particular useful at the moment, would eventually become more useful if/when lazy-loading for iframes is added to the specs. However removing the context makes it even less useful. Thinking that without context there's no point of having it at all at the moment.

> `wp_set_image_loading_attr` allows modifying the loading attribute for images specifically on an individual level.

Right. However there are some inconsistencies:
- This filter runs only in some cases, not in all of the cases where WP is adding the attribute.
- It is designed so plugins can add it to other places where adding the loading attribute makes sense.

Passing more context about why, where, how it runs would cover more cases and generally make it more useful. "Restricting" the context means that plugins won't be able to use it in certain cases.

Reply:
> Same as above. Less context means less useful hooks :)

Generally yes, but not really in this case IMO. It would help distinguish what these two filters should be used for. One would be for enabling/disabling the feature for a specific HTML element overall, the other would be for controlling per image. Removing `$context` from the more generic `wp_lazy_loading_enabled` filter would clarify its intent and not remove _any_ capability, because exactly that could be done with the more specific `wp_set_image_loading_attr` filter still.

> It is not particularly useful at the moment, would eventually become more useful if/when lazy-loading for iframes is added to the specs.

I somewhat agree with that. Given that the other filter currently provides more capabilities, how about we remove the generic `wp_lazy_loading_enabled` filter for now? We could consider adding it back once more HTML elements start supporting `loading`.

Reply:
> Removing $context from the more generic `wp_lazy_loading_enabled` filter would clarify its intent and not remove any capability, because exactly that could be done with the more specific `wp_set_image_loading_attr` filter still.

Yes, ideally that should have been the case. Unfortunately `wp_set_image_loading_attr` is fired only when adding the attribute to HTML content. If we remove the context from `wp_lazy_loading_enabled` how would you know when it was called from `get_avatar` or from `wp_get_attachment_image`?